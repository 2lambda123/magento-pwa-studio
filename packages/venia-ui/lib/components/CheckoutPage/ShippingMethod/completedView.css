.root {
    height: 100%;
}

.button {
    height: 100%;
}

.container {
    display: grid;
    grid-template-rows: auto 1fr;
    height: 100%;
    row-gap: 1em;
}

.contents {
    display: grid;
    grid-template-rows: auto auto;
    margin-top: 1rem;
    row-gap: 0.5rem;
    text-align: left;
}

.editButton {
    display: grid;
    grid-template-columns: auto auto;
    column-gap: 0.5rem;
}

.editButtonText {
    text-decoration: underline;
}

.error {
    align-self: center;
    color: rgb(var(--venia-error));
}

.free {
    text-transform: uppercase;
}

.heading {
    font-weight: 600;
}

.titleContainer {
    display: flex;
    justify-content: space-between;
    width: 100%;
}

/*
 *  Mobile-specific styles.
 */
@media (max-width: 960px) {
    .editButtonText {
        display: none;
    }
}
