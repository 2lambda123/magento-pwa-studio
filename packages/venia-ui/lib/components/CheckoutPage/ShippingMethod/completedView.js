import React from 'react';
import { arrayOf, bool, func, number, shape, string } from 'prop-types';
import { Edit2 as EditIcon } from 'react-feather';

import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../../classify';
import Icon from '../../Icon';
import defaultClasses from './completedView.css';

const Done = props => {
    const { selectedShippingMethod, shippingMethods, showUpdateMode } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    let contents;
    if (!selectedShippingMethod) {
        // Error state.
        contents = (
            <span className={classes.error}>
                {`Error loading selected shipping method. Please select again.`}
            </span>
        );
    } else {
        const selectedShippingMethodObject = shippingMethods.find(method => {
            return selectedShippingMethod === method.serializedValue;
        });
        const { amount, method_title } = selectedShippingMethodObject;
        const { currency, value } = amount;

        const priceElement = value ? (
            <div>
                <Price value={value} currencyCode={currency} />
            </div>
        ) : (
            <span className={classes.free}>Free</span>
        );

        contents = (
            <div className={classes.contents}>
                <span>{method_title}</span>
                {priceElement}
            </div>
        );
    }

    return (
        <div className={classes.root}>
            <div className={classes.container}>
                <span className={classes.titleContainer}>
                    <h5 className={classes.heading}>Shipping Method</h5>
                    <button className={classes.button} onClick={showUpdateMode}>
                        <div className={classes.editButton}>
                            <Icon size={18} src={EditIcon} />
                            <span className={classes.editButtonText}>
                                {'Edit'}
                            </span>
                        </div>
                    </button>
                </span>
                {contents}
            </div>
        </div>
    );
};

export default Done;

Done.propTypes = {
    classes: shape({
        button: string,
        container: string,
        contents: string,
        editButton: string,
        editButtonText: string,
        error: string,
        free: string,
        heading: string,
        root: string,
        titleContainer: string
    }),
    selectedShippingMethod: string,
    shippingMethods: arrayOf(
        shape({
            amount: shape({
                currency: string,
                value: number
            }),
            available: bool,
            carrier_code: string,
            carrier_title: string,
            method_code: string,
            method_title: string,
            serializedValue: string.isRequired
        })
    ),
    showUpdateMode: func
};
