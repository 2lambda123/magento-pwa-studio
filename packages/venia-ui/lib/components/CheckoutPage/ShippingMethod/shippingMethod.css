.root {
    border: none;
    border-radius: 0px;
    border-bottom: 1px solid rgb(var(--venia-border));
    min-height: 10rem;
    padding: 1.5rem 0rem;

    display: grid;
    grid-template-rows: auto 1fr;
    row-gap: 1rem;
}

.done {
    composes: root;
    border: 1px solid rgb(var(--venia-border));
    border-radius: 5px;
    padding: 1.5rem 1rem;
}

.editingHeading {
    text-transform: uppercase;
}

.form {
    display: grid;
    grid-template-rows: 1fr auto;
    row-gap: 1.5rem;
}

.formButtons {
    display: flex;
    justify-content: center;
}
