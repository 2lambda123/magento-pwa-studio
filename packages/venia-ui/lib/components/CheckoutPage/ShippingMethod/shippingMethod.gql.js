import gql from 'graphql-tag';

import { PriceSummaryFragment } from '@magento/venia-ui/lib/components/CartPage/PriceSummary/priceSummaryFragments';

import {
    AvailableShippingMethodsFragment,
    SelectedShippingMethodFragment
} from './shippingMethodFragments.gql';

export const GET_SELECTED_AND_AVAILABLE_SHIPPING_METHODS = gql`
    query getSelectedAndAvailableShippingMethods($cartId: String!) {
        cart(cart_id: $cartId) @connection(key: "Cart") {
            id
            ...AvailableShippingMethodsFragment
            ...SelectedShippingMethodFragment
        }
    }
    ${AvailableShippingMethodsFragment}
    ${SelectedShippingMethodFragment}
`;

export const SET_SHIPPING_METHOD = gql`
    mutation SetShippingMethod(
        $cartId: String!
        $shippingMethod: ShippingMethodInput!
    ) {
        setShippingMethodsOnCart(
            input: { cart_id: $cartId, shipping_methods: [$shippingMethod] }
        ) @connection(key: "setShippingMethodsOnCart") {
            cart {
                id
                ...SelectedShippingMethodFragment
                ...PriceSummaryFragment
                # Intentionally do not re-fetch available methods because
                #  a) they are wrong in the mutation response
                #  b) it is expensive to recalculate.
            }
        }
    }
    ${SelectedShippingMethodFragment}
    ${PriceSummaryFragment}
`;

export default {
    mutations: {
        setShippingMethod: SET_SHIPPING_METHOD
    },
    queries: {
        getSelectedAndAvailableShippingMethods: GET_SELECTED_AND_AVAILABLE_SHIPPING_METHODS
    }
};
