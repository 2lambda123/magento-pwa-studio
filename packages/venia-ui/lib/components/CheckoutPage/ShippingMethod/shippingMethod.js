import React, { Fragment } from 'react';
import { bool, func, shape, string } from 'prop-types';
import { Form } from 'informed';

import {
    displayStates,
    useShippingMethod
} from '@magento/peregrine/lib/talons/CheckoutPage/useShippingMethod';

import { mergeClasses } from '../../../classify';
import Button from '../../Button';
import CompletedView from './completedView';
import ShippingRadios from './shippingRadios';
import UpdateModal from './updateModal';
import defaultClasses from './shippingMethod.css';

import {
    GET_SHIPPING_METHODS,
    GET_SELECTED_SHIPPING_METHOD,
    SET_SHIPPING_METHOD
} from './shippingMethod.gql';

const ShippingMethod = props => {
    const { onSave, pageIsUpdating, setPageIsUpdating } = props;

    const talonProps = useShippingMethod({
        mutations: {
            setShippingMethod: SET_SHIPPING_METHOD
        },
        onSave,
        queries: {
            getShippingMethods: GET_SHIPPING_METHODS,
            getSelectedShippingMethod: GET_SELECTED_SHIPPING_METHOD
        },
        setPageIsUpdating
    });

    const {
        displayState,
        handleCancelUpdate,
        handleSubmit,
        isLoadingShippingMethods,
        isLoadingSelectedShippingMethod,
        isUpdateMode,
        selectedShippingMethod,
        shippingMethods,
        showUpdateMode
    } = talonProps;

    const isLoading =
        isLoadingSelectedShippingMethod || isLoadingShippingMethods;

    const classes = mergeClasses(defaultClasses, props.classes);

    let contents;
    switch (displayState) {
        case displayStates.EDITING:
            {
                const continueDisabled =
                    pageIsUpdating || !shippingMethods.length;

                contents = (
                    <div className={classes.root}>
                        <h3 className={classes.editingHeading}>
                            {'Shipping Method'}
                        </h3>
                        <Form className={classes.form} onSubmit={handleSubmit}>
                            <ShippingRadios
                                isLoading={isLoading}
                                selectedShippingMethod={selectedShippingMethod}
                                shippingMethods={shippingMethods}
                            />
                            {!isLoading && (
                                <div className={classes.formButtons}>
                                    <Button
                                        priority="normal"
                                        type="submit"
                                        disabled={continueDisabled}
                                    >
                                        {'Continue to Payment Information'}
                                    </Button>
                                </div>
                            )}
                        </Form>
                    </div>
                );
            }
            break;

        case displayStates.DONE:
        default: {
            contents = (
                <div className={classes.done}>
                    <CompletedView
                        selectedShippingMethod={selectedShippingMethod}
                        shippingMethods={shippingMethods}
                        showUpdateMode={showUpdateMode}
                    />
                </div>
            );
        }
    }

    return (
        <Fragment>
            {contents}
            <UpdateModal
                handleCancel={handleCancelUpdate}
                handleSubmit={handleSubmit}
                isLoading={isLoading}
                isOpen={isUpdateMode}
                pageIsUpdating={pageIsUpdating}
                selectedShippingMethod={selectedShippingMethod}
                shippingMethods={shippingMethods}
            />
        </Fragment>
    );
};

ShippingMethod.propTypes = {
    classes: shape({
        done: string,
        editingHeading: string,
        form: string,
        formButtons: string,
        root: string
    }),
    onSave: func.isRequired,
    pageIsUpdating: bool,
    setPageIsUpdating: func.isRequired
};

export default ShippingMethod;
