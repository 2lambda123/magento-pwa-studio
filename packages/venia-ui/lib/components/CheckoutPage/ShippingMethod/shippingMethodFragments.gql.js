import gql from 'graphql-tag';

/**
 * WARNING: This fragment can cause slowdowns on the network request because it
 * causes the backend to make remote queries to carriers for real-time data.
 * Make sure you are conscious about the side effects when including it in queries and mutations.
 */
export const AvailableShippingMethodsFragment = gql`
    fragment AvailableShippingMethodsFragment on Cart {
        id
        shipping_addresses {
            available_shipping_methods {
                amount {
                    currency
                    value
                }
                available
                carrier_code
                carrier_title
                method_code
                method_title
            }
        }
    }
`;

export const SelectedShippingMethodFragment = gql`
    fragment SelectedShippingMethodFragment on Cart {
        id
        shipping_addresses {
            selected_shipping_method {
                amount {
                    currency
                    value
                }
                carrier_code
                method_code
                method_title
            }
        }
    }
`;

export const ShippingMethodsFragment = gql`
    fragment ShippingMethodsFragment on Cart {
        id
        ...AvailableShippingMethodsFragment
        ...SelectedShippingMethodFragment
        shipping_addresses {
            country {
                code
            }
            postcode
            region {
                code
            }
        }
    }
    ${AvailableShippingMethodsFragment}
    ${SelectedShippingMethodFragment}
`;
