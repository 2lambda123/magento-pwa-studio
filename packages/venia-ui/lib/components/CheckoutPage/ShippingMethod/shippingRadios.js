import React from 'react';
import { arrayOf, bool, number, shape, string } from 'prop-types';

import { mergeClasses } from '../../../classify';
import LoadingIndicator from '../../LoadingIndicator';
import RadioGroup from '../../RadioGroup';
import ShippingRadio from '../../CartPage/PriceAdjustments/ShippingMethods/shippingRadio';
import defaultClasses from './shippingRadios.css';

const ERROR_MESSAGE =
    'Error loading shipping methods. Please ensure a shipping address is set and try again.';

const ShippingRadios = props => {
    const { isLoading, selectedShippingMethod, shippingMethods } = props;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (isLoading) {
        return (
            <LoadingIndicator classes={{ root: classes.loadingRoot }}>
                {'Loading shipping methods...'}
            </LoadingIndicator>
        );
    }

    if (!shippingMethods.length) {
        return <span className={classes.error}>{ERROR_MESSAGE}</span>;
    }

    const radioGroupClasses = {
        message: classes.radioMessage,
        radioLabel: classes.radioLabel,
        root: classes.radioRoot
    };

    const shippingRadios = shippingMethods.map(method => {
        const label = (
            <ShippingRadio
                currency={method.amount.currency}
                name={method.method_title}
                price={method.amount.value}
            />
        );
        const value = method.serializedValue;

        return { label, value };
    });

    return (
        <RadioGroup
            classes={radioGroupClasses}
            field="shipping_method"
            initialValue={selectedShippingMethod}
            items={shippingRadios}
        />
    );
};

export default ShippingRadios;

ShippingRadios.propTypes = {
    classes: shape({
        loadingRoot: string,
        radioMessage: string,
        radioLabel: string,
        radioRoot: string
    }),
    isLoading: bool,
    selectedShippingMethod: string,
    shippingMethods: arrayOf(
        shape({
            amount: shape({
                currency: string,
                value: number
            }),
            available: bool,
            carrier_code: string,
            carrier_title: string,
            method_code: string,
            method_title: string,
            serializedValue: string.isRequired
        })
    )
};
