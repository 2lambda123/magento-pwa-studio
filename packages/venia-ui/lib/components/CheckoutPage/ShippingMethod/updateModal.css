.root {
    /* The modal takes up the entire screen. */
    position: fixed;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;

    /* AppContext takes care of "graying out" the background. */
    background-color: transparent;
    z-index: 3;

    /*
     * The modal is a grid container with its contents
     * horizontally and vertically centered.
     */
    display: grid;
    grid-template-rows: 1fr auto 1fr;
    grid-template-columns: 1fr auto 1fr;
    grid-template-areas:
        '. ........ .'
        '. contents .'
        '. ........ .';

    /* The modal starts hidden. */
    opacity: 0;
    visibility: hidden;
    transition-duration: 192ms;
    transition-timing-function: var(--venia-anim-out);
    transition-property: opacity, visibility;
}

.root_open {
    composes: root;

    box-shadow: 1px 0 rgb(var(--venia-border));

    /* The modal animates to being visible. */
    opacity: 1;
    visibility: visible;
    transition-duration: 224ms;
    transition-timing-function: var(--venia-anim-in);
}

.contents {
    /* This is the only grid item in the modal. */
    grid-area: contents;
    background-color: white;

    /* It sets maximum sizes so its body can handle overflow. */
    height: 100%;
    min-height: 416px; /* 16:9 aspect ratio. */
    max-height: 90vh;
    width: 740px;
    /* But nothing is allowed to overflow contents itself. */
    overflow: hidden;

    /* It is itself a grid container for its children. */
    display: grid;
    grid-template-rows: auto 1fr auto;
    grid-template-areas:
        'header'
        'body'
        'footer';
}

/*
 * Header and children.
 */

.header {
    grid-area: header;

    border-bottom: 1px solid rgb(var(--venia-border));
    display: grid;
    grid-template-columns: auto auto;
    padding: 0.875rem;
}

.headerButton {
    justify-self: right;
}

.headerText {
    align-self: center;
    color: rgb(var(--venia-text-alt));
}

/*
 * Body and children.
 */

.body {
    grid-area: body;

    padding: 1rem;
    overflow: auto;
}

/* 
 * Footer and children.
 */

.footer {
    grid-area: footer;
    padding: 2.5rem;

    display: grid;
    grid-template-columns: auto auto;
    column-gap: 2rem;
}

.footerCancelButton {
    composes: root_normalPriority from '../../Button/button.css';
    justify-self: right;
    width: 2rem;
}

.footerSubmitButton {
    composes: root_highPriority from '../../Button/button.css';
    width: 2rem;
}

/*
 * Mobile-specific styles.
 *
 * Instead of being a full-page modal,
 * display within the confines of the right drawer.
 */

@media (max-width: 960px) {
    .root {
        top: 0;
        bottom: 0;
        height: 100%;

        left: auto;
        right: 0;
        max-width: 360px;

        /* In mobile, there's only the contents grid item. */
        grid-template-rows: 1fr;
        grid-template-columns: 1fr;
        grid-template-areas: 'contents';

        transform: translate3d(100%, 0, 0);
        transition-property: opacity, transform, visibility;
    }

    .root_open {
        composes: root;

        transform: translate3d(0, 0, 0);
    }

    .contents {
        max-height: 100%;
        width: 100%;
    }
}
