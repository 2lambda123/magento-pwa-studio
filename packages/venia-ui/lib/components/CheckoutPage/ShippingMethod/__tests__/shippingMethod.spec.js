import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import {
    displayStates,
    useShippingMethod
} from '@magento/peregrine/lib/talons/CheckoutPage/useShippingMethod';

import ShippingMethod from '../shippingMethod';

jest.mock('../../../../classify');
jest.mock(
    '@magento/peregrine/lib/talons/CheckoutPage/useShippingMethod',
    () => {
        const useShippingMethodTalon = jest.requireActual(
            '@magento/peregrine/lib/talons/CheckoutPage/useShippingMethod'
        );
        const spy = jest.spyOn(useShippingMethodTalon, 'useShippingMethod');

        return Object.assign(useShippingMethodTalon, {
            useShippingMethod: spy
        });
    }
);
jest.mock('../completedView', () => 'Completed View Component');
jest.mock('../updateModal', () => 'Update Modal Component');

const talonProps = {
    displayState: displayStates.EDITING,
    handleCancelUpdate: jest.fn(),
    handleSubmit: jest.fn(),
    isLoadingShippingMethods: false,
    isLoadingSelectedShippingMethod: false,
    isUpdateMode: false,
    selectedShippingMethod: 'flatrate|flatrate',
    shippingMethods: [
        {
            amount: {
                currency: 'USD',
                value: 99
            },
            available: true,
            carrier_code: 'flatrate',
            carrier_title: 'Flat Rate',
            method_code: 'flatrate',
            method_title: 'Flat Rate',
            serializedValue: 'flatrate|flatrate'
        }
    ],
    showUpdateMode: jest.fn()
};

test('it renders correctly', () => {
    // Arrange.
    useShippingMethod.mockReturnValueOnce(talonProps);

    // Act.
    const instance = createTestInstance(
        <ShippingMethod
            pageIsUpdating={true}
            onSave={jest.fn()}
            setPageIsUpdating={jest.fn()}
        />
    );

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('it renders correctly in done mode', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        displayState: displayStates.DONE
    };
    useShippingMethod.mockReturnValueOnce(myTalonProps);

    // Act.
    const instance = createTestInstance(
        <ShippingMethod
            pageIsUpdating={true}
            onSave={jest.fn()}
            setPageIsUpdating={jest.fn()}
        />
    );

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});
