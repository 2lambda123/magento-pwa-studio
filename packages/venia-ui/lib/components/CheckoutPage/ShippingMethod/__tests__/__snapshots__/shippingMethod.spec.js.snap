// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
Array [
  <div
    className="root"
  >
    <h3
      className="editingHeading"
    >
      Shipping Method
    </h3>
    <form
      className="form"
      onKeyDown={[Function]}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="root"
      >
        <label
          className="root"
        >
          <input
            checked={true}
            className="input"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="flatrate|flatrate"
          />
          <span
            className="label"
          >
            <span>
              Flat Rate
            </span>
            <div
              className="price"
            >
              <div />
            </div>
          </span>
        </label>
      </div>
      <p
        className="root"
      />
      <div
        className="formButtons"
      >
        <button
          className="root_normalPriority"
          disabled={true}
          type="submit"
        >
          <span
            className="content"
          >
            Continue to Payment Information
          </span>
        </button>
      </div>
    </form>
  </div>,
  <Update Modal Component
    handleCancel={[MockFunction]}
    handleSubmit={[MockFunction]}
    isLoading={false}
    isOpen={false}
    pageIsUpdating={true}
    selectedShippingMethod="flatrate|flatrate"
    shippingMethods={
      Array [
        Object {
          "amount": Object {
            "currency": "USD",
            "value": 99,
          },
          "available": true,
          "carrier_code": "flatrate",
          "carrier_title": "Flat Rate",
          "method_code": "flatrate",
          "method_title": "Flat Rate",
          "serializedValue": "flatrate|flatrate",
        },
      ]
    }
  />,
]
`;

exports[`it renders correctly in done mode 1`] = `
Array [
  <div
    className="done"
  >
    <Completed View Component
      selectedShippingMethod="flatrate|flatrate"
      shippingMethods={
        Array [
          Object {
            "amount": Object {
              "currency": "USD",
              "value": 99,
            },
            "available": true,
            "carrier_code": "flatrate",
            "carrier_title": "Flat Rate",
            "method_code": "flatrate",
            "method_title": "Flat Rate",
            "serializedValue": "flatrate|flatrate",
          },
        ]
      }
      showUpdateMode={[MockFunction]}
    />
  </div>,
  <Update Modal Component
    handleCancel={[MockFunction]}
    handleSubmit={[MockFunction]}
    isLoading={false}
    isOpen={false}
    pageIsUpdating={true}
    selectedShippingMethod="flatrate|flatrate"
    shippingMethods={
      Array [
        Object {
          "amount": Object {
            "currency": "USD",
            "value": 99,
          },
          "available": true,
          "carrier_code": "flatrate",
          "carrier_title": "Flat Rate",
          "method_code": "flatrate",
          "method_title": "Flat Rate",
          "serializedValue": "flatrate|flatrate",
        },
      ]
    }
  />,
]
`;
